clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -1960,9 +1961,11 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
                     llvm::MDNode::get(VMContext, argBaseTypeNames));
     Fn->setMetadata("kernel_arg_type_qual",
                     llvm::MDNode::get(VMContext, argTypeQuals));
+    // FIXME: What is the purpose of this metadata for ESIMD? Can we
+    // reuse kernel_arg_exclusive_ptr or kernel_arg_runtime_aligned ?
```

Comment:
What is the FIXME for?

Reply:
The metadata generated for ESIMD is `kernel_arg_accessor_ptr`. I was wondering if backend can use the new metadata `kernel_arg_exclusive_ptr` instead (currently under email discussion)

Reply:
`kernel_arg_accessor_ptr` is a list of boolean values 1 per argument, which tells whether the corresponding argument originates from an accessor or a USM pointer. `kernel_arg_exclusive_ptr` AFAICT has a very different semantics.
To avoid too many MD kinds, the common denominator could be `kernel_arg_ptr_desc` or `kernel_arg_ptr_info` which is able to represent both kinds of information - needed by ESIMD and this patch. Per-argument value could be e.g. an integer encoding needed flags or something similar.

Reply:
Both kernel_arg_exclusive_ptr and kernel_arg_accessor_ptr are a list of boolean values 1 per argument, which at the moment is true for base pointer of accessor and false otherwise. In fact, this list of booleans is maintained in the same SmallVector `argSYCLAccessorPtrs` for both metadata. 

The semantics for both `kernel_arg_accessor_ptr `and `kernel_arg_exclusive_ptr `(from an implementation perspective) are the same. My only concern is, at some point in the future `kernel_arg_exclusive_ptr `will extend to involve other pointers in certain situations. The spec says "kernel_arg_exclusive_ptr metadata guarantees that the kernel pointer argument, or pointers that derive from it, will not be dereferenced outside current invocation of the kernel" USM pointers do not meet this guarantee and therefore will be false. If ESIMD is differentiating between USM and accessor pointer for a similar reason we could possibly get rid of `kernel_arg_accessor_ptr `and use `kernel_arg_exclusive_ptr `

Reply:
Based on email discussion, I have removed the FIXME here - https://github.com/intel/llvm/pull/6351